chain/bitcoind.go

```diff
@@ -0,0 +1,1050 @@
+
+	c.onBlockConnected(&blockHash, height,
+		block.Header.Timestamp)
+
+	return relevantTxs, nil
+}
+
+// filterTx filters a single transaction against the client's watch list.
+func (c *BitcoindClient) filterTx(tx *wire.MsgTx,
+	blockDetails *btcjson.BlockDetails, notify bool) (bool,
+	*wtxmgr.TxRecord, error) {
+
+	txDetails := btcutil.NewTx(tx)
+	if blockDetails != nil {
+		txDetails.SetIndex(blockDetails.Index)
+	}
+
+	rec, err := wtxmgr.NewTxRecordFromMsgTx(txDetails.MsgTx(), time.Now())
+	if err != nil {
+		log.Errorf("Cannot create transaction record for relevant "+
+			"tx: %v", err)
+		return false, nil, err
+	}
+	if blockDetails != nil {
+		rec.Received = time.Unix(blockDetails.Time, 0)
+	}
+
+	var notifyTx bool
+
+	// If we already know this is a relevant tx from a previous ntfn, we
+	// can shortcut the filter process and let the caller know the filter
+	// matches.
+	if _, ok := c.memPool[tx.TxHash()]; ok {
+		return true, rec, nil
+	}
+
+	// Cycle through outputs and check if we've matched a known address.
+	// Add any matched outpoints to watchOutPoints.
+	for i, out := range tx.TxOut {
+		_, addrs, _, err := txscript.ExtractPkScriptAddrs(
+			out.PkScript, c.chainParams)
+		if err != nil {
+			log.Warnf("Couldn't parse output script in %s:%d: %v",
+				tx.TxHash(), i, err)
+			continue
+		}
+		for _, addr := range addrs {
+			if _, ok := c.watchAddrs[addr.EncodeAddress()]; ok {
+				notifyTx = true
+				c.watchOutPoints[wire.OutPoint{
+					Hash:  tx.TxHash(),
+					Index: uint32(i),
+				}] = struct{}{}
+			}
+		}
+	}
+
+	// If an output hasn't already matched, see if an input will.
+	if !notifyTx {
+		for _, in := range tx.TxIn {
+			if _, ok := c.watchOutPoints[in.PreviousOutPoint]; ok {
+				notifyTx = true
+				break
+			}
+		}
+	}
+
+	// If we have a match and it's not mined, notify the TX. If the TX is
+	// mined, we notify as part of FilteredBlockConnected. The boolean map
+	// value will let us know if we last saw it as mined or unmined.
+	if notifyTx {
+		c.memPool[tx.TxHash()] = struct{}{}
+		if blockDetails == nil {
+			c.onRelevantTx(rec, nil)
+		}
+	}
+
+	return notifyTx, rec, nil
+}
+
+// handler maintains a queue of notifications and the current state (best
+// block) of the chain.
+func (c *BitcoindClient) handler() {
+	hash, height, err := c.GetBestBlock()
+	if err != nil {
+		log.Errorf("Failed to receive best block from chain server: %v", err)
+		c.Stop()
+		c.wg.Done()
+		return
+	}
+
+	bs := &waddrmgr.BlockStamp{Hash: *hash, Height: height}
+
+	// TODO: Rather than leaving this as an unbounded queue for all types of
+	// notifications, try dropping ones where a later enqueued notification
+	// can fully invalidate one waiting to be processed.  For example,
+	// blockconnected notifications for greater block heights can remove the
+	// need to process earlier blockconnected notifications still waiting
+	// here.
+
+	var notifications []interface{}
```

Comment:
Can you add a TODO here to factor this logic out into a shared data structure? It's currently repeated a few times between the various implementations of the `chain` backend.

Reply:
TODO added.